src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ProvideCorrectArgumentsToFormattingMethods.cs

```diff
@@ -22,64 +24,105 @@ public class ProvideCorrectArgumentsToFormattingMethodsAnalyzer : DiagnosticAnal
+                    var invocation = (IInvocationOperation)context.Operation;
 
-                    StringFormatInfo.Info? info = formatInfo.TryGet(invocation.TargetMethod, operationContext);
+                    StringFormatInfo.Info? info = formatInfo.TryGet(invocation.TargetMethod, context);
                     if (info == null || invocation.Arguments.Length <= info.FormatStringIndex)
                     {
                         // not a target method
                         return;
                     }
 
                     IArgumentOperation formatStringArgument = invocation.Arguments[info.FormatStringIndex];
-                    if (!object.Equals(formatStringArgument?.Value?.Type, formatInfo.String) ||
+                    if (!Equals(formatStringArgument?.Value?.Type, formatInfo.String) ||
                         !(formatStringArgument?.Value?.ConstantValue.Value is string))
                     {
                         // wrong argument
                         return;
                     }
 
                     var stringFormat = (string)formatStringArgument.Value.ConstantValue.Value;
-                    int expectedStringFormatArgumentCount = GetFormattingArguments(stringFormat);
+                    var stringFormatIndexes = GetStringFormatItemIndexes(stringFormat);
+                    int expectedStringFormatArgumentCount = stringFormatIndexes?.Count ?? -1;
 
-                    // explicit parameter case
-                    if (info.ExpectedStringFormatArgumentCount >= 0)
+                    // Validate string format item indexes
+                    if (expectedStringFormatArgumentCount > 0)
                     {
-                        // __arglist is not supported here
-                        if (invocation.TargetMethod.IsVararg)
+                        var missingIndexes = Enumerable.Range(0, stringFormatIndexes.Max())
+                            .Except(stringFormatIndexes);
+                        if (missingIndexes.Any())
                         {
-                            // can't deal with this for now.
+                            context.ReportDiagnostic(invocation.CreateDiagnostic(MissingFormatItemRule));
+
+                            // There are some missing format indexes, we don't know if they are just missing
+                            // or if the highest numbers have been shifted so skip the arguments actually
+                            // provided to the invocation.
                             return;
                         }
+                    }
 
+                    // __arglist is not supported here (see https://github.com/dotnet/roslyn/issues/7346)
+                    // but we want to skip it only in C# since VB doesn't support __arglist
+                    if (info.ExpectedStringFormatArgumentCount >= 0 &&
+                        invocation.TargetMethod.IsVararg &&
+                        invocation.Language == LanguageNames.CSharp)
```

Comment:
This filter was suggested but not present in the original implementation. Let me know if you prefer to revert this.

Reply:
I am fine with this change